### ISSUE TITLE
Never allow the cursor to be behind the popup (forcing it to the bottom)

### ISSUE THREAD

**Issue by dominikduda** [wellle/context.vim#42]
nvim info:
```
NVIM v0.5.0-dev
Build type: RelWithDebInfo
Lua 5.1
Compilation: /usr/bin/cc -g -O2 -fdebug-prefix-map=/build/neovim-LDECaw/neovim-0.5.0+ubuntu1+git202001312020-14a8b3b-00e710e=. -fstack-protector-strong -Wformat -Werror=format-security -U_FORTIFY_SOURCE -D_FORTIFY_SOURCE=1 -O2 -g -Og -g -Wall -Wextra -pedantic -Wno-unused-parameter -Wstrict-prototypes -std=gnu99 -Wshadow -Wconversion -Wmissing-prototypes -Wimplicit-fallthrough -Wvla -fstack-protector-strong -fdiagnostics-color=auto -DINCLUDE_GENERATED_DECLARATIONS -D_GNU_SOURCE -DNVIM_MSGPACK_HAS_FLOAT32 -DNVIM_UNIBI_HAS_VAR_FROM -DMIN_LOG_LEVEL=3 -I/build/neovim-LDECaw/neovim-0.5.0+ubuntu1+git202001312020-14a8b3b-00e710e/build/config -I/build/neovim-LDECaw/neovim-0.5.0+ubuntu1+git202001312020-14a8b3b-00e710e/src -I/build/neovim-LDECaw/neovim-0.5.0+ubuntu1+git202001312020-14a8b3b-00e710e/.deps/usr/include -I/usr/include -I/build/neovim-LDECaw/neovim-0.5.0+ubuntu1+git202001312020-14a8b3b-00e710e/build/src/nvim/auto -I/build/neovim-LDECaw/neovim-0.5.0+ubuntu1+git202001312020-14a8b3b-00e710e/build/include
Compiled by buildd@lcy01-amd64-015

Features: +acl +iconv +tui
See ":help feature-compile"

   system vimrc file: "$VIM/sysinit.vim"
  fall-back for $VIM: "/usr/share/nvim"

Run :checkhealth for more info
```

Plugin config:
```
let g:context_highlight_border = 'LineNr'
let g:context_border_char = '⬍'
```

Presentation: https://webmshare.com/play/YzDKO

**Comment by wellle** [comment]
Hey, thanks for the report, but this is actually intended behavior. The idea is that if the cursor gets moved to the top we want to move the context popup somewhere else, so it doesn't obstruct where the cursor is.

There are some mappings to avoid that to happen, but there are still many ways to trigger this case.

There's a big comment at the top of [autoload/context/mapping.vim](https://github.com/wellle/context.vim/blob/master/autoload/context/mapping.vim) describing that current state.

Does that make sense?

**Comment by dominikduda** [comment]
Yes, now I understand why it made that way.

I personally think though that much more intuitive use case would be to never allow the cursor to move behind it.

I'd prefer it to behave more like when it's set to `let g:context_presenter = 'preview'` but this causes other problems.

Could the following option be leveraged (by setting it on the fly to context window height) to add such behavior?
```
    " Always show at least 1 line above/below the cursor
        set scrolloff=1
```
This could be optional.

This prevents cursor to ever move closer than one line from the top/bottom of the screen. 
For example, when you go to the previous search result which is at top of the screen and `set scrolloff` is set to 1, the window will scroll and the cursor will jump to the result (now in the second line).

**Comment by wellle** [comment]
Hey, yes I fully agree. It would be great to make sure the cursor never moves behind the popup. And it's still a goal for me to improve on that. It's a bit complicated though and I just wanted the popup feature out before investing too much time in this.

I did actually try using `scrolloff` for this. One downside is that it applies to the bottom of the screen too, which is annoying because it would be different depending on the height of the popup window, so commands like `L` would behave unpredictably.

More importantly `scrolloff` has other weirdness as you try to scroll up the screen. The problem is that we need to change the `scrolloff` setting dynamically every time the height of the popup changes. And this can push the cursor down. So there are cases where your cursor is stuck while you keep pressing `k` for example.

So as far as I know the best way forward would be as described in the code comment I mentioned. And I might work on this at some point.

**Comment by dominikduda** [comment]
@wellle Lets call the amount of squares between cursor row (in relation to the top of the screen) and popup last row N.

We should be able to get this N before the popup is actually drawn.
Then we could could do `feedkeys("\<C-y>", "tx")` N times.
Would this fix the issue?

**Comment by wellle** [comment]
Yes, that's kind what I do in the custom mappings. But now consider the case where the update has been triggered by `<C-E>`. If now you apply `<C-Y>` to fix the cursor position, this negates the `<C-E>`. So in that case we need to do `j` instead of `<C-Y>`. That's what I was trying to say in that code comment. If there was a scroll motion we need to fix by moving the cursor, if there was a cursor move we need to fix by scrolling. But both of these motion groups (moving cursor and scrolling) consist of quite a few Vim commands. So we need a non trivial amount of custom mappings to make this work.

**Comment by dominikduda** [comment]
I understand the problem now.

I think what would fix it is manipulation of scrolloff value (as on the attached webm file).

Changing it to context height (4 in the example) then back to original value leaves no changes to any config and works universally correctly for both scrolling and cursor moving options.

This could be used at the same moment but instead of `feedkeys` in the comment above.

https://webmshare.com/play/6VY1V

**Comment by wellle** [comment]
As for the `scrolloff` idea, I resurrected my earlier attempt and applied it to the current master. I pushed it to branch `42-scrolloff`. Please have a look.

It seems fairly good for the most part. As mentioned above, some commands like `zb` and `L` behave unpredictably because of the varying scrolloff setting. But what's more, try scrolling though a file with varying context popup window heights with `k` and `<C-E>`. I do feel like there's some weirdness going on. It seems like increasing the scrolloff value has some delay before it actually pushes the cursor down. Specifically, when you scroll up (by using `k`) and that results in a bigger context, the context still gets moved down temporarily. And if you scroll with `<C-E>` slowly, you can get stuck in such cases. (if that would move the cursor up into the preview window the increased scrolloff value just pushes the cursor done again, negating the scroll).

So all in all it doesn't seem good enough to me. Let me know how it feels to you 👍

**Comment by dominikduda** [comment]
It's much better for me now. I only update it on cursor hold though.

I have one more idea:
What if on cursor move we would store last cursor line.
Would this create a universal way to distinguish between scrolling screen and scrolling by cursor?
What I mean we could compare the last cursor line with current.

**Comment by wellle** [comment]
Hmm, I think I had considered that briefly but felt like it had some flaw. Thinking about it a bit more I feel like that might actually work. Thanks for the suggestion, I'll play around with it 👍

**Comment by wellle** [comment]
@dominikduda: I worked on this on and off and finally merged #65 to implement this idea. I think it turned out pretty well, I've been using it for a few days now and it's so much better now with the context always at the top. Please try it out!

And thanks again for opening this issue, I think this change makes this plugin so much more clean! ❤️
